### ISSUE TITLE
ModifiedControlVariableCheck: NPE when SuppressWarnings is used

### ISSUE THREAD

**Issue by clausfod** [checkstyle/checkstyle#2327]
Test case:
```
/var/tmp $ cat 3.java 
package test;
public class Test {
    public String foo(String request) {
        List<String> list;
        for (String eventDataType : list) {

            String eventDataHandler = "";
            if (eventDataHandler != null) {
                try {
                    @SuppressWarnings(value = "unchecked")
                    String eventData;

                } catch (Exception e) {
                    // no smth
                }

            }
        }

        return "";
    }
}

/var/tmp $ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="ModifiedControlVariableCheck"/>
  </module>
</module>

/var/tmp $ java -jar checkstyle-6.11.2-all.jar -c config.xml   3.java
Exception in thread "main" java.lang.NullPointerException
	at com.puppycrawl.tools.checkstyle.checks.coding.ModifiedControlVariableCheck.checkIdent(ModifiedControlVariableCheck.java:278)
	at com.puppycrawl.tools.checkstyle.checks.coding.ModifiedControlVariableCheck.visitToken(ModifiedControlVariableCheck.java:199)
	at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:388)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:499)
	at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:330)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:203)
	at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:79)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:265)
	at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:277)
	at com.puppycrawl.tools.checkstyle.Main.main(Main.java:118)
```

**Comment by mkordas** [comment]
@clausfod, Can you also provide stacktrace?

**Comment by clausfod** [comment]
There is no stacktrace - only this: Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin: 2.16:checkstyle (default-cli) on project xxxx.xxx: Execution default-cli of goal org.apache.maven.plugins:maven-checkstyle-plugin:2.16 :checkstyle failed. NullPointerException -> [Help 1]

**Comment by mkordas** [comment]
You need to add "-e" as parameter.

**Comment by clausfod** [comment]
´´´´xml
project xxx: Execution default-cli of goal org.a
pache.maven.plugins:maven-checkstyle-plugin:2.16:checkstyle failed.
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor
.java:224)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor
.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor
.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProje
ct(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProje
ct(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThre
adedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(Lifecycl
eStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:862)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:286)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:197)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.
java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAcces
sorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Laun
cher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.jav
a:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(La
uncher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:
356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-c
li of goal org.apache.maven.plugins:maven-checkstyle-plugin:2.16:checkstyle fail
ed.
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(Default
BuildPluginManager.java:145)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor
.java:208)
        ... 20 more
Caused by: java.lang.NullPointerException
        at com.puppycrawl.tools.checkstyle.checks.coding.ModifiedControlVariable
Check.checkIdent(ModifiedControlVariableCheck.java:278)
        at com.puppycrawl.tools.checkstyle.checks.coding.ModifiedControlVariable
Check.visitToken(ModifiedControlVariableCheck.java:199)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.jav
a:388)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.jav
a:499)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:330)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker
.java:203)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(Abst
ractFileSetCheck.java:79)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:265)
        at org.apache.maven.plugin.checkstyle.exec.DefaultCheckstyleExecutor.exe
cuteCheckstyle(DefaultCheckstyleExecutor.java:252)
        at org.apache.maven.plugin.checkstyle.AbstractCheckstyleReport.executeRe
port(AbstractCheckstyleReport.java:473)
        at org.apache.maven.plugin.checkstyle.CheckstyleReport.executeReport(Che
ckstyleReport.java:154)
        at org.apache.maven.reporting.AbstractMavenReport.generate(AbstractMaven
Report.java:255)
        at org.apache.maven.reporting.AbstractMavenReport.execute(AbstractMavenR
eport.java:143)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(Default
BuildPluginManager.java:134)
        ... 21 more
´´´´

**Comment by romani** [comment]
please prove that issue exists by our CLI on latest released version (6.11.2) to remove maven plugin from consideration(maven plugin use old version - https://maven.apache.org/plugins/maven-checkstyle-plugin/history.html). We need config and input file from you to reproduce a problem.

Example of perfectly reported issue - https://github.com/checkstyle/checkstyle/issues/2250 .

**Comment by clausfod** [comment]
Hi,

Thanks for trying to help solving the problem with Checkstyle. I can't attach a specific java file, because I don't know which one of the 10.000+ java classes in the "src" folder has the problem.

You can overwrite the checkstyle maven plugin to use another version by doing this (as stated in the initally comment):

```xml
<plugin>
<groupId>org.apache.maven.plugins</groupId>
<artifactId>maven-checkstyle-plugin</artifactId>
<version>2.16</version>
<configuration>
..
</configuration>
<dependencies>
<dependency>
<groupId>com.puppycrawl.tools</groupId>
<artifactId>checkstyle</artifactId>
<version>${checkstyle.version}</version>
</dependency>
</dependencies>
</plugin>
```

So the checkstyle maven plugin fails with 6.11.2 of Checkstyle. I have reproduced the error when runing Checktstyle from the command line (which indicate it is a problem with Checkstyle analysis or configuration of the tool):

```xml
java -jar checkstyle-6.11.2-all.jar -c checkstyle_no_suppression.xml src/
Exception in thread "main" java.lang.NullPointerException
        at com.puppycrawl.tools.checkstyle.checks.coding.ModifiedControlVariable
Check.checkIdent(ModifiedControlVariableCheck.java:278)
        at com.puppycrawl.tools.checkstyle.checks.coding.ModifiedControlVariable
Check.visitToken(ModifiedControlVariableCheck.java:199)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.jav
a:388)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.jav
a:499)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:330)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker
.java:203)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(Abst
ractFileSetCheck.java:79)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:265)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:277)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:118)
```

The analysis is against the "src" folder which have around 10.000 java classes. So which one should I attach (the stacktrace doesn't tell) ?

Here is the xml file:
```xml

<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://jee-qa/jee-qa/trunk/checkstyle/configuration_1_2.dtd">

<module name="Checker">

    <!-- Checks whether files end with a new line. -->
    <module name="NewlineAtEndOfFile">
        <property name="lineSeparator" value="lf" />
    </module>

    <!-- Checks for size violations -->
    <module name="FileLength">
        <property name="max" value="1500" />
    </module>
    <module name="FileTabCharacter" />

    <module name="Translation" />

    <module name="TreeWalker">
        <property name="cacheFile" value="target/checkstyle-cachefile" />

        <!-- Check for Javadocs Comments ? -->
        <module name="JavadocType">
            <property name="scope" value="public" />
            <property name="tokens" value="INTERFACE_DEF, CLASS_DEF" />
        </module>

        <!-- Checks for Naming Conventions. -->
        <module name="ConstantName" />
        <module name="LocalVariableName" />
        <module name="LocalFinalVariableName" />
        <module name="MemberName" />
        <module name="MethodName" />
        <module name="PackageName" />
        <module name="ParameterName" />
        <module name="StaticVariableName" />
        <module name="TypeName" />

        <!-- Checks for imports -->
        <module name="AvoidStaticImport" />
        <module name="IllegalImport" />
        <module name="RedundantImport" />
        <module name="UnusedImports" />
        <module name="AvoidStarImport">
            <property name="excludes" value="java.*, javax.*" />
        </module>
        <module name="ImportOrder">
            <property name="groups" value="java, javax, *" />
            <property name="ordered" value="true" />
            <property name="separated" value="true" />
            <property name="option" value="above" />
        </module>

        <!-- Checks for whitespace -->
        <module name="MethodParamPad" />
        <module name="NoWhitespaceAfter" />
        <module name="NoWhitespaceBefore" />
        <module name="ParenPad" />
        <module name="TypecastParenPad" />
        <module name="WhitespaceAfter" />
        <module name="WhitespaceAround" />
        <module name="Indentation" />
        <module name="GenericWhitespace" />

        <!-- Modifier Checks -->
        <module name="ModifierOrder" />
        <module name="RedundantModifier" />

        <!-- Checks for blocks. You know, those {}'s -->
        <module name="EmptyBlock">
            <property name="option" value="text" />
        </module>
        <module name="LeftCurly" />
        <module name="RightCurly" />
        <module name="AvoidNestedBlocks" />

        <!-- Checks for common coding problems -->
        <module name="CovariantEquals" />
        <module name="EmptyStatement" />
        <module name="EqualsAvoidNull" />
        <module name="EqualsHashCode" />

        <!-- module name="FinalLocalVariable" / -->
        <module name="IllegalInstantiation" />
        <module name="HiddenField">
            <property name="tokens" value="VARIABLE_DEF" />
        </module>
        <module name="MissingSwitchDefault" />
        <module name="InnerAssignment" />
        <module name="MagicNumber">
            <property name="tokens" value="NUM_DOUBLE, NUM_FLOAT" />
            <property name="ignoreNumbers" value="0, 0.5, 0.75, 1" />
        </module>
        <module name="ModifiedControlVariable" />
        <module name="SimplifyBooleanExpression" />
        <module name="SimplifyBooleanReturn" />
        <module name="StringLiteralEquality" />
        <module name="NestedTryDepth">
            <property name="max" value="3" />
        </module>
        <module name="IllegalCatch">
            <property name="illegalClassNames"
                value="java.lang.Throwable, java.lang.Error" />
        </module>
        <module name="IllegalThrows" />
        <module name="PackageDeclaration" />
        <module name="MultipleStringLiterals">
            <property name="allowedDuplicates" value="3" />
            <property name="ignoreStringsRegexp" value=".{0,12}" />
        </module>
        <module name="MultipleVariableDeclarations" />

        <module name="SuperClone" />
        
        <module name="DeclarationOrder" />
        <module name="ExplicitInitialization" />
        <module name="DefaultComesLast" />
        <module name="FallThrough" />

        <!-- Checks for class design -->
        <module name="HideUtilityClassConstructor" />
        <module name="NeedBraces">
            <property name="tokens"
                value="LITERAL_DO, LITERAL_FOR, LITERAL_WHILE" />
        </module>
        <module name="ThrowsCount">
            <property name="max" value="3" />
        </module>

        <!-- complexity checks -->
        <module name="BooleanExpressionComplexity" />
        <module name="ClassDataAbstractionCoupling">
            <property name="max" value="15" />
        </module>
        <module name="ClassFanOutComplexity">
            <property name="max" value="30" />
        </module>
        <module name="CyclomaticComplexity">
            <property name="max" value="20" />
        </module>
        <module name="NPathComplexity">
            <property name="max" value="350" />
        </module>
        <module name="JavaNCSS">
            <property name="methodMaximum" value="150" />
        </module>
        <module name="ParameterNumber">
            <property name="max" value="7" />
        </module>

        <!-- Miscellaneous other checks. -->
        <module name="Regexp">
            <property name="format" value="System\.(out|err)\.println" />
            <property name="illegalPattern" value="true" />
            <property name="message"
                value="System.out.println() - never write to the console - use logging instead" />
        </module>
        <module name="Regexp">
            <property name="format" value="\.printStackTrace\(\)" />
            <property name="illegalPattern" value="true" />
            <property name="message"
                value=" printStackTrace() never write to the console - use logging instead" />
        </module>
        <module name="Regexp">
            <property name="format" value="Class\.forName" />
            <property name="illegalPattern" value="true" />
            <property name="message"
                value=" do you want to use Class.forName - use context classloader instead" />
        </module>
        <module name="Regexp">
            <property name="format" value="Place description here" />
            <property name="illegalPattern" value="true" />
            <property name="message"
                value=" Place description here - is an autogenerated comment - are noise - remove it" />
        </module>
        <module name="Regexp">
            <property name="format"
                value="Header: |Revision: |@version|@since" />
            <property name="illegalPattern" value="true" />
            <property name="message"
                value="Header: |Revision: |@version|@since - are noise - remove it" />
        </module>
        <module name="Regexp">
            <property name="format"
                value="User:|Date:|Time:|Created by IntelliJ IDEA.|Copyright " />
            <property name="illegalPattern" value="true" />
            <property name="message"
                value=" User:|Date:|Time:|Created by IntelliJ IDEA.|Copyright are autogenerated comments - are noise - remove it" />
        </module>
        <module name="Regexp">
            <property name="format" value=" Cloneable" />
            <property name="illegalPattern" value="true" />
            <property name="message"
                value="Cloneable is broken by design - use copy constructors instead" />
        </module>

        <module name="ArrayTypeStyle" />
        <module name="UncommentedMain" />
        <module name="UpperEll" />

        <module name="LineLength">
            <property name="max" value="144" />
        </module>
        <module name="MethodLength">
            <property name="countEmpty" value="false" />
            <property name="max" value="50" />
        </module>

        <module name="AnonInnerLength">
            <property name="max" value="40" />
        </module>
    </module>
</module>
```

**Comment by MEZk** [comment]
@romani 
See https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java#L278

When [ast token](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java#L276) does not have children than identAST is equal to null. And we have NPE on line 278.
We should find test input for this situation and/or check whether identAST is equal to null on line 278.

**Comment by romani** [comment]
@clausfod , sorry that we can not help you quickly, we have registered issue for printing problematic file in logs.

To let you catch a problem please remove from config all Checks except for ModifiedControlVariable, and try to remove launch CLI on subfolders of your "src" folder, validation will be quick, by means of this approach you will step down to problematic lowest package quickly, file you can find by running of CLI on certain file or set of files.

Sorry for caused inconvenience and manual work to detect a problem.

**Comment by clausfod** [comment]
Here is the code with the problem:
```java
    public EfpiEvent createEvent(AdviseringOmHaendelseModtagRequestType request) {
            List<EfpiEventType> eventDataTypeList = eventTypeListMap.get(event.getEventRequest().getEventDataType());
                for (EfpiEventType eventDataType : eventDataTypeList) {

                    AbstractEventDataHandler eventDataHandler = eventDataHandlerMap.get(eventDataType);
                    if (eventDataHandler != null) {
                        try {
                            @SuppressWarnings(value = "unchecked")
                            EfpiEventData eventData = eventDataHandler.readEventData(event, eventDataType, haendelseNoegledata);
                            event.addEventData(eventData);

                        } catch (AccessDeniedException ade) {
                            LOGGER.debug(ade.getMessage(), ade);
                            event.setEventStatus(EfpiEventStatus.IGNORE);

                        } catch (Exception e) {
                            LOGGER.error(e.getMessage(), e);
                            event.setEventStatus(EfpiEventStatus.ERROR_IMPORT);
                        }

                    } else {
                        LOGGER.warn("Unable to find EventDataHandler based on event: " + event);
                    }
            }

        return event;
    }
```

If I remove the @SuppressWarnings(value = "unchecked") in the for loop the check is running fine... I guess you need to implement what @MEZk mentioned earlier in this issue.

Kind regards
Claus

**Comment by romani** [comment]
released in 6.12
